src/rules/no_await_in_loop.rs

```diff
@@ -0,0 +1,428 @@
+    } else {
+      FunctionVisitor::new(&self.root_visitor, false)
+    };
+    func_visitor.visit_arrow_expr(arrow_expr, parent);
+  }
+
+  fn visit_await_expr(&mut self, await_expr: &AwaitExpr, parent: &dyn Node) {
+    self.root_visitor.add_diagnostic(await_expr.span);
+    swc_ecma_visit::visit_await_expr(self, await_expr, parent);
+  }
+}
+
+struct FunctionVisitor<'a> {
+  root_visitor: &'a NoAwaitInLoopVisitor,
+  is_async: bool,
+}
+
+impl<'a> FunctionVisitor<'a> {
+  fn new(root_visitor: &'a NoAwaitInLoopVisitor, is_async: bool) -> Self {
+    Self {
+      root_visitor,
+      is_async,
+    }
+  }
+}
+
+impl<'a> Visit for FunctionVisitor<'a> {
+  fn visit_function(&mut self, func: &Function, parent: &dyn Node) {
+    let mut func_visitor =
+      FunctionVisitor::new(&self.root_visitor, func.is_async);
+    swc_ecma_visit::visit_function(&mut func_visitor, func, parent);
+  }
+
+  fn visit_arrow_expr(&mut self, arrow_expr: &ArrowExpr, parent: &dyn Node) {
+    let mut func_visitor =
+      FunctionVisitor::new(&self.root_visitor, arrow_expr.is_async);
+    swc_ecma_visit::visit_arrow_expr(&mut func_visitor, arrow_expr, parent);
+  }
+
+  fn visit_for_stmt(&mut self, for_stmt: &ForStmt, parent: &dyn Node) {
+    if self.is_async {
+      let mut loop_visitor = LoopVisitor::new(&self.root_visitor);
+      loop_visitor.visit_for_stmt(for_stmt, parent);
+    } else {
+      swc_ecma_visit::visit_for_stmt(self, for_stmt, parent);
+    }
+  }
+
+  fn visit_for_of_stmt(&mut self, for_of_stmt: &ForOfStmt, parent: &dyn Node) {
+    if self.is_async && for_of_stmt.await_token.is_none() {
+      let mut loop_visitor = LoopVisitor::new(&self.root_visitor);
+      loop_visitor.visit_for_of_stmt(for_of_stmt, parent);
+    } else {
+      swc_ecma_visit::visit_for_of_stmt(self, for_of_stmt, parent);
+    }
+  }
+
+  fn visit_for_in_stmt(&mut self, for_in_stmt: &ForInStmt, parent: &dyn Node) {
+    if self.is_async {
+      let mut loop_visitor = LoopVisitor::new(&self.root_visitor);
+      loop_visitor.visit_for_in_stmt(for_in_stmt, parent);
+    } else {
+      swc_ecma_visit::visit_for_in_stmt(self, for_in_stmt, parent);
+    }
+  }
+
+  fn visit_while_stmt(&mut self, while_stmt: &WhileStmt, parent: &dyn Node) {
+    if self.is_async {
+      let mut loop_visitor = LoopVisitor::new(&self.root_visitor);
+      loop_visitor.visit_while_stmt(while_stmt, parent);
+    } else {
+      swc_ecma_visit::visit_while_stmt(self, while_stmt, parent);
+    }
+  }
+
+  fn visit_do_while_stmt(
+    &mut self,
+    do_while_stmt: &DoWhileStmt,
+    parent: &dyn Node,
+  ) {
+    if self.is_async {
+      let mut loop_visitor = LoopVisitor::new(&self.root_visitor);
+      loop_visitor.visit_do_while_stmt(do_while_stmt, parent);
+    } else {
+      swc_ecma_visit::visit_do_while_stmt(self, do_while_stmt, parent);
+    }
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::test_util::*;
+
+  #[test]
+  fn no_await_in_loop_valid() {
```

Comment:
can you add this source to valid tests, please?
```ts
async function myLoop() {
  for (let i = await something(); i < n; i++) {}
}
```
This is considered an error,
That shouldn't be the case.

Reply:
Thank you for the review!
Oh, I missed that. I'll fix it.